backend/server/src/nix/mod.rs

```diff
@@ -36,30 +37,32 @@ impl EvalService {
         }
     }
 
-    pub fn run(self) {
-        tokio::spawn(async {
-            self.listen().await;
-        });
-    }
-
-    async fn listen(mut self) {
-        loop {
-            match self.drv_receiver.recv().await {
-                Some(EvalTask::Job(drv)) => {
-                    if let Err(e) = self.run_nix_eval_jobs(drv.file_path).await {
-                        warn!("Ran into error when query eval job: {}", e);
-                    };
-                }
-                Some(EvalTask::TraverseDrv(drv)) => {
-                    if let Err(e) = self.traverse_drvs(&drv).await {
-                        warn!("Ran into error when query drv information: {}", e);
-                    }
-                }
+    pub async fn run(mut self, cancellation_token: CancellationToken) {
+        while let Some(request) = cancellation_token
+            .run_until_cancelled(self.drv_receiver.recv())
+            .await
+        {
+            let task = match request {
+                Some(task) => task,
                 None => {
-                    warn!("Eval reciever channel shutdown");
+                    warn!("Eval receiver channel closed, shutting down");
+                    break;
                 }
+            };
+
+            let result = match task {
+                EvalTask::Job(drv) => self.run_nix_eval_jobs(drv.file_path).await,
+                EvalTask::TraverseDrv(drv) => self.traverse_drvs(&drv).await,
+            };
+
+            if let Err(e) = result {
+                error!(error = %e, "Failed to handle task")
             }
         }
+
+        self.db_service.close().await;
```

Comment:
I'd rather move this to the `main` function, after we waited for all other services to shutdown. Having it here feels a bit out of place.